File: Algorithms/0838_Push_Dominoes/Java/PushDominoesSolution1.java

// Space: O(1)
// Time: O(n)

class Solution {
        public String pushDominoes(String dominoes) {
        int[] cache = new int[dominoes.length()];
        int force = 0;

        for (int i = 0; i < dominoes.length(); i++) {
            if (dominoes.charAt(i) == 'R') force = cache.length;
            else if (dominoes.charAt(i) == 'L') force = 0;
            else force = Math.max(force - 1, 0);
            cache[i] += force;
        }

        force = 0;
        for (int i = dominoes.length() - 1; i >= 0; i--) {
            if (dominoes.charAt(i) == 'L') force = cache.length;
            else if (dominoes.charAt(i) == 'R') force = 0;
            else force = Math.max(force - 1, 0);
            cache[i] -= force;
        }

        StringBuilder res = new StringBuilder();
        for (int i : cache) {
            if (i > 0) res.append("R");
            else if (i < 0) res.append("L");
            else res.append(".");
        }
        return res.toString();
        }
}

File: Algorithms/0838_Push_Dominoes/Python/Push_Dominoes_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        length = len(dominoes)
        cache = [0 for _ in range(length)]

        force = 0
        for i in range(length):
            if dominoes[i] == 'R': force = length
            elif dominoes[i] == 'L': force = 0
            else: force = max(force - 1, 0)
            cache[i] += force

        force = 0
        for i in range(len(cache) - 1, -1, -1):
            if dominoes[i] == 'L': force = length
            elif dominoes[i] == 'R': force = 0
            else: force = max(force - 1, 0)
            cache[i] -= force

        res = ''
        for i in cache:
            if i > 0: res += 'R'
            elif i < 0: res += 'L'
            else: res += '.'

        return res




